src/rust/storage/seg/src/ttl_buckets/ttl_bucket.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+//! TTL bucket containing a segment chain which stores items with a similar TTL
+//! in an ordered fashion.
+//!
+//! TTL Bucket:
+//! ```text
+//! ┌──────────────┬──────────────┬─────────────┬──────────────┐
+//! │   HEAD SEG   │   TAIL SEG   │     TTL     │   SEGMENTS   │
+//! │              │              │             │              │
+//! │    32 bit    │    32 bit    │    32 bit   │    32 bit    │
+//! ├──────────────┼──────────────┴─────────────┴──────────────┤
+//! │  NEXT MERGE  │                  PADDING                  │
+//! │              │                                           │
+//! │    32 bit    │                  96 bit                   │
+//! ├──────────────┴───────────────────────────────────────────┤
+//! │                         PADDING                          │
+//! │                                                          │
+//! │                         128 bit                          │
+//! ├──────────────────────────────────────────────────────────┤
+//! │                         PADDING                          │
+//! │                                                          │
+//! │                         128 bit                          │
+//! └──────────────────────────────────────────────────────────┘
+//! ```
+
+use crate::*;
+use core::num::NonZeroU32;
+
+/// Each ttl bucket contains a segment chain to store items with a similar TTL
+/// in an ordered fashion. The first segment to expire will be the head of the
+/// segment chain. This allows us to efficiently scan across the [`TtlBuckets`]
+/// and expire segments in an eager fashion.
+pub struct TtlBucket {
+    head: Option<NonZeroU32>,
+    tail: Option<NonZeroU32>,
+    ttl: i32,
+    segments: i32,
+    next_to_merge: Option<NonZeroU32>,
+    _pad: [u8; 44],
+}
+
+#[cfg(feature = "dump")]
+#[derive(Serialize, Deserialize)]
+pub struct TtlBucketDump {
+    ttl: i32,
+    head: i32,
+}
+
+impl TtlBucket {
+    pub(super) fn new(ttl: i32) -> Self {
+        Self {
+            head: None,
+            tail: None,
+            ttl,
+            segments: 0,
+            next_to_merge: None,
+            _pad: [0; 44],
+        }
+    }
+
+    pub fn head(&self) -> Option<NonZeroU32> {
+        self.head
+    }
+
+    pub fn set_head(&mut self, id: Option<NonZeroU32>) {
+        self.head = id;
+    }
+
+    pub fn next_to_merge(&self) -> Option<NonZeroU32> {
+        self.next_to_merge
+    }
+
+    pub fn set_next_to_merge(&mut self, next: Option<NonZeroU32>) {
+        self.next_to_merge = next;
+    }
+
+    // expire segments from this TtlBucket, returns the number of segments expired
+    pub(super) fn expire(&mut self, hashtable: &mut HashTable, segments: &mut Segments) -> usize {
```

Comment:
Can we add some (verbose) logging to track the time it takes to expire all the qualified segments? This could be set at a verbose level that's the same or one higher than the overall event (expire for the global `ttl_buckets`). I'm slightly concerned that the worst case is a noticeable glitch for very large heaps (e.g. if we flush a full heap) and want to at least have an easy way to clock that if needed.

Reply:
Good idea! Should really be a stat too.

Reply:
Added a new counter and a debug level log message, LMK if this is good now.